For review

xosmig opened the pull request.

xosmig committed:
Add class diagram

xosmig committed:
Implement most on environment package. Update the class diagram respectively

xosmig committed:
Create Command and CommandExecutor interfaces

xosmig committed:
Add token class and token expansion

xosmig committed:
Add tokenizer

xosmig committed:
Add parsing

xosmig committed:
Fix bugs in AssignmentParser, add LocalAssignmentCommand

xosmig committed:
Implement StandardCommandExecutor

xosmig committed:
I don't have the time to start from scratch :(

xosmig committed:
Debug and test class `Tokenizer`

xosmig committed:
Add `PipeToken` and `SemicolonToken`, fix escaping rules in `Tokenizer`

xosmig committed:
Rewrite and test `AssignmentParser`

xosmig committed:
Test and debug `PipeParser`

xosmig committed:
Test and debug `SemicolonParser`

xosmig committed:
Change the approach to escaping symbols

Previously `Tokenizer` consumed all the backslashes and it was impossible to understand whether a character was escaped or not in later stages of parsing.
Now I introduced class `CmdChar`, which is basicaly a char with a boolean flag, whether it is escaped or not.
I also introduced class `CmdString` and `CmdStringBuilder` which are prety much like the classes from the standard library, but `CmdString` consists of `CmdChar`s.

xosmig committed:
Implement variable expansion

xosmig committed:
Finish basic implementation

xosmig committed:
Initial commit

xosmig committed:
Merge branch 'master' of github.com:xosmig/aush

xosmig committed:
Make the class diagram and the code constistent with each other

xosmig committed:
Add run.sh for easy running the shell on Unix systems

xosmig committed:
Add class diagram to the README

xosmig committed:
Fix a little mistake in the class diagram

xosmig committed:
Add ci settings

xosmig committed:
Fix tests compilation error

xosmig committed:
Add travis ci badge to the README

xosmig committed:
Document the `Environment` class and remove all its modifying methods

`Environment.Builder` should be used for all `Environment` modifications.

xosmig committed:
Document `commands` package along with small refactoring

xosmig committed:
Document and refactor token package

xosmig committed:
Refactor and document the rest of the code

yurii-litvinov commented on src/main/java/com/xosmig/swdesignhw/aush/commands/executor/StandardCommandExecutor.java:
```diff
@@ -0,0 +1,167 @@
+            try {
+                env.getOutput().doRedirection(process.getInputStream());
+            } catch (Exception e) {
+                // oops
```
Мм. Хоть что-то сделать нужно, так оно вообще все исключения проглатиывать будет. Если ничего не должно бросать исключения внутри, лучше выкинуть RuntimeException, если должно, то ловить что-то конкретное

yurii-litvinov commented on src/main/java/com/xosmig/swdesignhw/aush/environment/Inherit.java:
```diff
@@ -0,0 +1,48 @@
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public final class Inherit {
```
Я бы сказал, что этим классам тоже нужна документация

yurii-litvinov replied:
Напоминаю

yurii-litvinov commented on src/main/java/com/xosmig/swdesignhw/aush/textui/TextUserInterface.java:
```diff
@@ -0,0 +1,84 @@
+    // For dependency injection
+    public CommandExecutor executor = new StandardCommandExecutor();
+    public TextCommandCompiler compiler = new BashLikeCommandCompiler();
+    public Environment environment;
```
Для dependency injection не лучше ли использовать конструктор или в крайнем случае геттеры-сеттеры?

yurii-litvinov reviewed (changes requested):
В целом всё круто, но хочется больше встроенных команд. Например, в винде echo реализована как встроенная команда шелла, поэтому попытка вызывать её как внешний процесс заканчивается неуспехом. Хочется echo, wc, cat как встроенные команды, чтобы несчастные пользователи Windows тоже могли пользоваться этим шеллом.

xosmig committed:
Add echo, cat and wc builtins

yurii-litvinov commented on src/main/java/com/xosmig/swdesignhw/aush/ShellInternalError.java:
```diff
@@ -0,0 +1,23 @@
+package com.xosmig.swdesignhw.aush;
+
+public class ShellInternalError extends RuntimeException {
```
Я бы сказал, что этому классу тоже надо комментарий.

xosmig replied:
Done.

yurii-litvinov commented on src/main/java/com/xosmig/swdesignhw/aush/utils/Reference.java:
```diff
@@ -0,0 +1,6 @@
+public final class Reference<T> {
+
+    public T obj = null;
+}
```
Класс явно не из мира Java :) В Java вроде как идеоматично использовать массивы из одного элемента для этих целей. Впрочем, это можно не править, мне так больше нравится.

xosmig replied:
Fixed.

yurii-litvinov commented on src/main/java/com/xosmig/swdesignhw/aush/utils/Utils.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.InterruptedIOException;
+import java.io.OutputStream;
+
+public final class Utils {
```
Тоже надо комментарии же

xosmig replied:
Done.

yurii-litvinov commented on src/test/java/com/xosmig/swdesignhw/aush/commands/executor/StandardCommandExecutorTest.java:
```diff
@@ -0,0 +1,68 @@
+        );
+        assertOutput("     13      69     779\n");
+    }
+}
```
Тестовое покрытие не впечатляет. Cat, Exit и Pwd не покрыты вообще.

yurii-litvinov commented on src/test/java/com/xosmig/swdesignhw/aush/commands/executor/StandardCommandExecutorTest.java:
```diff
@@ -0,0 +1,68 @@
+    @Test
+    public void testEchoHelloWorld() throws Exception {
+        compileAndRun(env, "echo hello, world!");
+        assertOutput("hello, world!\n");
```
Хардкодить переносы строк в Java идеологически неверно. У меня вот ни один тест не прошёл

xosmig replied:
Сделал тесты windows-совместимыми и добавил виндовый CI

yurii-litvinov reviewed (changes requested):
На команде `> echo 123 | wc` шелл повисает. На `> echo 123 | cat` не повисает, но ничего не делает. Ожидается, что cat выведет свой входной поток.

```
> echo '123$x'
'1231'
```
А должно быть `123$x`

Поправьте, пожалуйста.

xosmig committed:
Rename ShellInternalError to ShellInternalException and add a single line of documentation

xosmig committed:
Use more idiomatic arrays of size 1 instead of Reference class

xosmig committed:
Try to make unit tests windows-compatible

xosmig committed:
Add appveyor configuration

xosmig committed:
Avoid using non-English characters in string literals

xosmig committed:
Print info level logs in appveyor

xosmig committed:
Fix testWcPoem expected value

xosmig committed:
Fix testWcPoem on windows, but deleting unnecessary converting to os-specific string format

xosmig committed:
Add windows ci badge to README

xosmig committed:
Fix the bag with hang pipe and actually run the parts of a pipe in parallel

xosmig committed:
Merge branch 'master' of github.com:xosmig/aush

xosmig committed:
Document and refactor Input / Output utility of the environment package

xosmig committed:
Fix appveyor build by avoiding using platform-dependant tests. Add expected semantic for cat builtin without arguments

xosmig committed:
Add test for echo piped to wc

xosmig committed:
Add a test for pwd builtin

xosmig committed:
Add tests for MultipleCommands and Assign

xosmig committed:
Fix testPwd on windows platform

xosmig committed:
Refactor unit tests

xosmig committed:
Add a test for exit command and fix its semantics inside semicolon-separated list of commands

xosmig commented:
@yurii-litvinov в целом, все пофиксил, добавил немного тестов. А вот одинарные кавычки мне поддерживать не хотелось бы. Это обязательно?
Вообще, у баша довольно кривая и неинтуитивная семантика кавычек и экранирования, которую довольно сложно воспроизвести.

xosmig committed:
Small refactoring

xosmig committed:
Fix the bug when escaping backslashes were restored during the token expansion

xosmig committed:
Add support for single quotes in shell

xosmig committed:
Add --info flag to the test command in travis ci

xosmig commented:
@yurii-litvinov добавил одинарные кавычки, это оказалось не так страшно, как я почему-то думал.

xosmig committed:
Fix the bug with incorrect number of bytes counted by wc command in case of non-textual files

xosmig committed:
Improve pipes: increase buffer size, add comments, avoid thread leaks

Two user threads were licked for every call to pipe.
Among other problems, it caused the program to hang when exit is requested.

xosmig committed:
Add codecov report uploading

xosmig committed:
Add codecov badge to README

yurii-litvinov commented on src/main/java/com/xosmig/swdesignhw/aush/environment/CloseableInput.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * An {@code Input} object, which can be closed.
+ * It depends on the particular implementation whether a call to {@code close} required or not.
+ * Nobody is going to read this comment anyway.
```
:)

yurii-litvinov commented on src/main/java/com/xosmig/swdesignhw/aush/environment/Environment.java:
```diff
@@ -0,0 +1,327 @@
+        public Path workingDir;
+        public PMap<String, String> varValues;
+        public int lastExitCode;
+        public boolean shouldExit;
```
А зачем они public?

xosmig replied:
Хм, какой-то артефакт прошлого, спасибо.

yurii-litvinov reviewed (approved):
Да, всё теперь очень годно (правда, часть про парсеры так и осталась без комментариев). Зачтена.

xosmig committed:
Remove one easter egg from comments

xosmig committed:
Change visibility of unnecessarily public fields in Environment.Builder and make the class final
